runtime/catalog-karapace/src/main/java/io/aklivity/zilla/runtime/catalog/karapace/internal/KarapaceCatalogHandler.java

```diff
@@ -84,54 +88,77 @@ public KarapaceCatalogHandler(
         this.event = new KarapaceEventContext(context);
         this.catalogId = catalogId;
-        this.cachedSchemas = cache.cachedSchemas;
-        this.cachedSchemaIds = cache.cachedSchemaIds;
+        this.cachedSchemas = cache.schemas;
+        this.cachedSchemaIds = cache.schemaIds;
     }
 
     @Override
     public String resolve(
         int schemaId)
     {
         String schema = null;
-        if (schemas.containsKey(schemaId))
+        if (schemaId != NO_SCHEMA_ID)
         {
-            schema = schemas.get(schemaId);
-        }
-        else
-        {
-            CompletableFuture<String> future = cachedSchemas.get(schemaId);
-            if (future == null)
+            if (schemas.containsKey(schemaId))
             {
-                CompletableFuture<String> newFuture = new CompletableFuture<String>();
-                future = cachedSchemas.putIfAbsent(schemaId, newFuture);
-                if (future == null)
+                schema = schemas.get(schemaId);
+            }
+            else
+            {
+                AtomicInteger eventStatus = new AtomicInteger();
+                CompletableFuture<CachedSchema> newFuture = new CompletableFuture<>();
+                CompletableFuture<CachedSchema> existing = cachedSchemas.get(schemaId);
+                if (existing != null && existing.isDone())
+                {
+                    try
+                    {
+                        CachedSchema cachedSchema = existing.get();
+                        if (cachedSchema != null)
+                        {
+                            eventStatus = cachedSchema.event;
+                        }
+                    }
+                    catch (Throwable ex)
+                    {
+                        existing.completeExceptionally(ex);
+                    }
+                }
+                CompletableFuture<CachedSchema> future = cachedSchemas.merge(schemaId, newFuture, (v1, v2) ->
+                    v1.getNow(CachedSchema.IN_PROGRESS).schema == null ? v2 : v1);
+                if (future == newFuture)
                 {
                     try
                     {
                         String response = sendHttpRequest(MessageFormat.format(SCHEMA_PATH, schemaId));
-                        newFuture.complete(response != null ? request.resolveSchemaResponse(response) : null);
+                        if (response == null && eventStatus.getAndIncrement() == 0)
+                        {
+                            event.unretrievableSchemaId(catalogId, schemaId);
+                        }
+                        else if (response != null && eventStatus.getAndSet(0) > 0)
+                        {
+                            event.retrievableSchemaId(catalogId, schemaId);
+                        }
+                        newFuture.complete(new CachedSchema(response != null ? request.resolveSchemaResponse(response) : null,
+                            eventStatus));
```

Comment:
We are repeating the check for response == or != null several times here, which makes the code a bit harder to read.

Suggest we have one block
```
if (response == null)
{
    if (retryAttempts.getAndIncrement() == 0)
    {
        event.unretrievableSchemaId(catalogId, schemaId);
    }
    newFuture.complete(new CachedSchema(null, retryAttempts));
}
else
{
    if (retryAttempts.getAndSet(0) > 0)
    {
        event.retrievableSchemaId(catalogId, schemaId);
    }
    newFuture.complete(new CachedSchema(request.resolveSchemaResponse(response), retryAttempts));
}
```